src/main/java/org/apache/maven/shared/utils/cli/Commandline.java

```diff
@@ -241,15 +241,16 @@ public void addSystemEnvironment()
     public String[] getEnvironmentVariables()
     {
         addSystemEnvironment();
-        String[] environmentVars = new String[envVars.size()];
-        int i = 0;
+        List<String> environmentVars = new ArrayList<>();
         for ( String name : envVars.keySet() )
         {
             String value = envVars.get( name );
-            environmentVars[i] = name + "=" + value;
-            i++;
+            if ( value != null )
+            {
+                environmentVars.add( name + "=" + value );
+            }
         }
-        return environmentVars;
+        return environmentVars.toArray( new String[0] );
```

Comment:
We could speed up this code a very little bit (some nanoseconds though) if we use `new String[environmentVars.size()]` instead. It prevents dropping the empty array and creating a new one, hence reduces GC pressure.

Reply:
> We could speed up this code a very little bit (some nanoseconds though) if we use `new String[environmentVars.size()]` instead. It prevents dropping the empty array and creating a new one, hence reduces GC pressure.

It seems, even across Apache repositories (https://github.com/apache/kafka/pull/9750, https://github.com/apache/tika/pull/372), that using empty array exactly is popular now.

Reply:
From InteliJ inspection:

> In older Java versions using pre-sized array was recommended, as the reflection call which is necessary to create an array of proper size was quite slow. However since late updates of OpenJDK 6 this call was intrinsified, making the performance of the empty array version the same and sometimes even better, compared to the pre-sized version. Also passing pre-sized array is dangerous for a concurrent or synchronized collection as a data race is possible between the size and toArray call which may result in extra nulls at the end of the array, if the collection was concurrently shrunk during the operation.